File: LAB4/task_5/prisoners_dilemma.go
/*
Package prisoners_dilemma реализует игру "Дилемма заключенного".
Содержит три стратегии: всегда сотрудничать, всегда предавать и "око за око".
*/
package main

import (
	"fmt"
	"math/rand"
	"time"
)

type Strategy func(int, []bool, []bool) bool

type PrisonersDilemma struct {
	rounds    int
	strategyA Strategy
	strategyB Strategy
	choicesA  []bool
	choicesB  []bool
	scoreA    int
	scoreB    int
}

/**
 * @brief Создает новый экземпляр игры.
 * @param a Стратегия игрока A.
 * @param b Стратегия игрока B.
 */
func NewPrisonersDilemma(a, b Strategy) *PrisonersDilemma {
	rand.Seed(time.Now().UnixNano())
	return &PrisonersDilemma{
		rounds:    rand.Intn(101) + 100,
		strategyA: a,
		strategyB: b,
	}
}

/**
 * @brief Рассчитывает очки на основе выборов игроков.
 */
func (pd *PrisonersDilemma) calculateScores(a, b bool) {
	if !a && !b {
		pd.scoreA += 4
		pd.scoreB += 4
	} else if a && !b {
		pd.scoreB += 20
	} else if !a && b {
		pd.scoreA += 20
	} else {
		pd.scoreA += 24
		pd.scoreB += 24
	}
}

/**
 * @brief Запускает серию раундов игры.
 */
func (pd *PrisonersDilemma) Run() {
	for round := 0; round < pd.rounds; round++ {
		choiceA := pd.strategyA(round, pd.choicesA, pd.choicesB)
		choiceB := pd.strategyB(round, pd.choicesB, pd.choicesA)

		pd.choicesA = append(pd.choicesA, choiceA)
		pd.choicesB = append(pd.choicesB, choiceB)

		pd.calculateScores(choiceA, choiceB)
	}
}

// Примеры стратегий
func AlwaysCooperate(round int, self, enemy []bool) bool {
	return true
}

func AlwaysBetray(round int, self, enemy []bool) bool {
	return false
}

func TitForTat(round int, self, enemy []bool) bool {
	if round == 0 {
		return true
	}
	if len(enemy) == 0 {
		return true
	}
	return enemy[len(enemy)-1]
}

func main() {
	game1 := NewPrisonersDilemma(TitForTat, AlwaysCooperate)
	game1.Run()
	fmt.Printf("TitForTat vs Всегда сотрудничать: %d - %d\n", game1.scoreA, game1.scoreB)

	game2 := NewPrisonersDilemma(TitForTat, AlwaysBetray)
	game2.Run()
	fmt.Printf("TitForTat vs Всегда предавать: %d - %d\n", game2.scoreA, game2.scoreB)
}

File: LAB4/task_5/PrisonersDilemma.cpp
/**
 * @file PrisonersDilemma.cpp
 * @brief Реализация игры "Дилемма заключенного" с тремя стратегиями.
 */

#include <iostream>
#include <vector>
#include <functional>
#include <random>

using namespace std;

using Strategy = function<bool(int, const vector<bool>&, const vector<bool>&)>;

class PrisonersDilemma {
private:
    int rounds;
    Strategy strategyA;
    Strategy strategyB;
    vector<bool> choicesA;
    vector<bool> choicesB;
    int scoreA = 0;
    int scoreB = 0;

    /**
     * @brief Генерирует случайное количество раундов от 100 до 200.
     */
    int generateRounds() {
        random_device rd;
        mt19937 gen(rd());
        uniform_int_distribution<> distr(100, 200);
        return distr(gen);
    }

    /**
     * @brief Рассчитывает очки для игроков на основе их выборов.
     * @param a Выбор игрока A.
     * @param b Выбор игрока B.
     */
    void calculateScores(bool a, bool b) {
        if (!a && !b) { // Оба предали
            scoreA += 4;
            scoreB += 4;
        } else if (a && !b) { // A сотрудничал, B предал
            scoreB += 20;
        } else if (!a && b) { // A предал, B сотрудничал
            scoreA += 20;
        } else { // Оба сотрудничали
            scoreA += 24;
            scoreB += 24;
        }
    }

public:
    /**
     * @brief Конструктор класса.
     * @param a Стратегия игрока A.
     * @param b Стратегия игрока B.
     */
    PrisonersDilemma(Strategy a, Strategy b) 
        : strategyA(a), strategyB(b), rounds(generateRounds()) {}

    /**
     * @brief Запускает игру на заданное количество раундов.
     */
    void run() {
        for (int round = 0; round < rounds; ++round) {
            bool choiceA = strategyA(round, choicesA, choicesB);
            bool choiceB = strategyB(round, choicesB, choicesA);
            
            choicesA.push_back(choiceA);
            choicesB.push_back(choiceB);
            
            calculateScores(choiceA, choiceB);
        }
    }

    int getScoreA() const { return scoreA; }
    int getScoreB() const { return scoreB; }
};

// Примеры стратегий
bool alwaysCooperate(int, const vector<bool>&, const vector<bool>&) {
    return true;
}

bool alwaysBetray(int, const vector<bool>&, const vector<bool>&) {
    return false;
}

bool titForTat(int round, const vector<bool>& self, const vector<bool>& enemy) {
    if (round == 0) return true;
    return enemy.back();
}

int main() {
    PrisonersDilemma game1(titForTat, alwaysCooperate);
    game1.run();
    cout << "TitForTat vs Всегда сотрудничать: " 
         << game1.getScoreA() << " - " << game1.getScoreB() << endl;

    PrisonersDilemma game2(titForTat, alwaysBetray);
    game2.run();
    cout << "TitForTat vs Всегда предавать: " 
         << game2.getScoreA() << " - " << game2.getScoreB() << endl;

    return 0;
}
